README.md

```diff
@@ -1 +1,523 @@
+
+## 4.3 JSON Examples
+
+``` js
+{
+  "some-wasm": {
+    "with": "ipfs://bafkreicgincymfzzumcqcrpmquyildzqppa3s2gv3scfy2csaqpgqpp7ke",
+    "do": "wasm/run",
+    "inputs": {
+      "func": "calculate",
+      "args": [
+        1,
+        "hello world",
+        {"c": {"ucan/promise": ["/", "some-other-action"]}},
+        {"a": 1, "b": 2, "c": 3}
+      ]
+    },
+    "ipvm/config": {
+      "secret": false,
+      "check": {
+        "optimistic": 17, 
+        "referee": "did:key:zStEZpzSMtTt9k2vszgvCwF4fLQQSyA15W5AQ4z3AR6Bx4eFJ5crJFbuGxKmbma4"
+      }
+    }
+  }
+}
+```
+
+``` json
+{
+  "set-dns": {
+    "with": "dns://example.com?TYPE=TXT",
+    "do": "crud/update",
+    "inputs": { 
+      "value": "hello world"
+    },
+    "meta": {
+      "ipvm/config": {
+        "secret": false
+        "timeout": [500, "milli", "seconds"],
+        "verification": "attestation"
+      }
+    }
+  }
+}
+```
+
+``` json
+{
+  "run-docker": {
+    "with": "ipfs://bafkreidvq3uqoxcxr44q5qhgdk5zk6jvziipyxguirqa6tkh5z5wtpesva",
+    "do": "docker/run",
+    "inputs": {
+      "entrypoint": ["/", "home"],
+      "mounts": {"mnt": "/"},
+      "outputs": {"tmp": "/tmp"},
+      "env":  {"$HELLO": "world"},
+      "workdir":  "/work"
+    },
+    "meta": {
+      "ipvm/config": {
+        "secret": false,
+        "check": {
+          "optimistic": 2, 
+          "referee": "did:key:zStEZpzSMtTt9k2vszgvCwF4fLQQSyA15W5AQ4z3AR6Bx4eFJ5crJFbuGxKmbma4"
+        }
+      }
+    }
+  }
+}
+```
+
+``` json
+{
+  "pipelined-docker-promise": {
+    "with": {"ucan/promise": ["/", "previous_action"]},
+    "do": "docker/run",
+    "inputs": {
+      "entrypoint": ["/", "home"],
+      "mounts": {"mnt": "/"},
+      "outputs": {"tmp": "/tmp"},
+      "env":  {"$HELLO": "world"},
+      "workdir":  "/work"
+    },
+    "meta": {
+      "ipvm/config": {
+        "disk": "200GB"
+      }
+    }
+  }
+}
+```
+
+# 5 Exception Handler
+
+If present, the OPTIONAL `catch` field MUST be run in response to a `Task` returning on the `Failure` branch. The deterministic & pure Wasm module MUST take a `Failure` object as input, and MUST return data in the following shape:
+
+``` ipldsch
+type Handle union {
+  | Success "ok"     -- End task with Success object
+  | String  "rewire" -- Task name inside the current Workflow
+  | String  "msg"    -- Format the error message and panic
+} respresentation keyed
+```
+
+If the `msg` branch is returned, the invocation MUST immediately rethrow with the update message.
+
+Note that while IPVM MUST treat the pure tasks together as transactional. It is not possible to roll back any destructive effects that have already been run. As such, it is RECOMMENDED to have few (if any) tasks depend on the output of a destructive effect, so they can be scheduled at the end of the workflow.
```

Comment:
any thoughts on how to enforce this in any way?

Reply:
I've been thinking about destructive actions almost like a cut in Prolog. You lose transactional guarantees as soon as you use them. We may want to include a flag or something to set your expected properties like transactional, but that's also probably a big ask of most devs. Handling this in tooling with e.g. a warning like "hey, just so you know, we won't be able to roll back prior to this email if something goes wrong"